edk2toollib/uefi/wincert.py

```diff
@@ -81,77 +141,181 @@ def Write(self, fs):
+        # Account for back compat. Behave differently for file streams.
+        if hasattr(in_data, 'seek'):
+            self.CertData = in_data.read()
         else:
-            self.PopulateFromFileStream(filestream)
+            self.CertData = in_data
 
-    def AddCertData(self, fs):
-        if(self.CertData is not None):
-            raise Exception("Cert Data not 0")
-        self.CertData = memoryview(fs.read())
         self.Hdr_dwLength = self.Hdr_dwLength + len(self.CertData)
+
     #
     # Method to un-serialize from a filestream
     #
 
     def PopulateFromFileStream(self, fs):
-        if(fs is None):
-            raise Exception("Invalid File stream")
+        if fs is None:
+            raise ValueError
 
-        # only populate from file stream those parts that are complete in the file stream
-        offset = fs.tell()
-        fs.seek(0, 2)
-        end = fs.tell()
-        fs.seek(offset)
+        # Determine the end of the stream.
+        current = fs.tell()
+        end = fs.seek(0, io.SEEK_END)
+        fs.seek(current)
 
-        if((end - offset) < WinCertUefiGuid.STATIC_STRUCT_SIZE):  # size of the static header data
-            raise Exception("Invalid file stream size")
+        # Make sure that we can at least parse the size field.
+        field_string = "<I"
+        field_size = struct.calcsize(field_string)
+        if (end - current) < field_size:
+            raise ValueError
 
-        self.Hdr_dwLength = struct.unpack("=I", fs.read(4))[0]
-        self.Hdr_wRevision = struct.unpack("=H", fs.read(2))[0]
-        self.Hdr_wCertificateType = struct.unpack("=H", fs.read(2))[0]
-        self.CertType = uuid.UUID(bytes_le=fs.read(16))
-        self.CertData = None
+        # Parse the size field.
+        (buffer_size,) = struct.unpack(field_string, fs.read(field_size))
 
-        if((end - fs.tell()) < 1):
-            raise Exception("Invalid File stream. No data for signature cert data")
+        if (end - current) < buffer_size:
+            raise ValueError
 
-        if((end - fs.tell()) < (self.Hdr_dwLength - WinCertUefiGuid.STATIC_STRUCT_SIZE)):
-            raise Exception("Invalid file stream size ")
+        fs.seek(current)
+        object_buffer = fs.read(buffer_size)
 
-        self.CertData = memoryview(fs.read(self.Hdr_dwLength - WinCertUefiGuid.STATIC_STRUCT_SIZE))
+        return self.Decode(object_buffer)
 
     def Print(self):
-        print("WinCertUefiGuid")
-        print("  Hdr_dwLength:         0x%X" % self.Hdr_dwLength)
-        print("  Hdr_wRevision:        0x%X" % self.Hdr_wRevision)
-        print("  Hdr_wCertificateType: 0x%X" % self.Hdr_wCertificateType)
-        print("  CertType:             %s" % str(self.CertType))
-        print("  CertData:             ")
-        cdl = self.CertData.tolist()
-        PrintByteList(cdl)
+        self.DumpInfo()
+
+    def DumpInfo(self):
+        print('EFI_FIRMWARE_IMAGE_AUTHENTICATION.AuthInfo.Hdr.dwLength         = {dwLength:08X}'
```

Comment:
not right

Reply:
What are we talking about here?

Reply:
Ah... I see it.
Weird... wonder why that wasn't throwing exceptions all over the place.

Reply:
Wait... it's fine. The Encode was broken, though. What else are we talking about?